README.md
# Ubigraph

![Image of Ubigraph](http://holsee.github.io/images/exconfeu/erlubi_vanilla.png)

Ubigraph is a different way to create interactive user interfaces with the purpose of visualizing huge amounts of data. Its visualizations are simply stunning. This version is shipped in binary form as a standalone server that responds to requests using XML-RPC. This makes it easy to use from C, C++, Java, Ruby, Python, Perl, and other languages for which XML-RPC implementations are available. Since XML-RPC uses TCP-IP, the server (which visualizes the graph) can be run on a different machine/operating system than the client (which is manipulating the graph). It is also possible to have multiple clients updating the graph simultaneously. (Note that for clients to be on different machines from the server, firewalls must be configured to allow traffic on port 20738.)

# Why does this repository exist?

Ubigraph was developed by ubietylab, but the project has stopped receiving updates since 2008. Also it is not possible to download the latest version through the official website (http://ubietylab.net/ubigraph/content/Downloads/index.php).

I was able to find the latest version (alpha-0.2.4, June 2008) and thought that making it available here in github would be easier for other developers to use this amazing tool. Besides using I think that others can continue to contribute to the improvement and those who know how to bring new versions.

I hope ubietylab can look at this repository and keep contributing to this beautiful project.

# Using

```bash
$ ./bin/ubigraph_server
```

A black window will appear


Changelog
Version alpha-0.2.4, 23 June 2008 (r1951)
* Lots of performance improvements, and a new section in the user's guide
  about performance tuning.
* New edge attributes: arrow_position, arrow_length, arrow_radius, 
  arrow_reverse
* New edge stroke kind: "none", does not draw the edge, but still draws
  decorations (i.e. arrowhead and label)
* New examples: arrows.py, routing.py
* Bug fix: vertex "visible" attribute not handled properly

Version alpha-0.2.3, 29 May 2008
* Better Python API: examples/Python/ubigraph_example.py
* Reduced dependency on shared libraries

Version alpha-0.2.2, 28 May 2008
* XMLRPC callbacks on left double-click; see Python/callback.py
  and Python/callback_webcrawler.py
* New platforms supported: Mac OS X PowerPC, Ubuntu 6.06
* arrowheads are back (will be more customizable in future)
* New API methods: new_vertex_style_w_id(s,parent) and 
  new_edge_style_w_id(s,parent)
* Performance and numeric-stability improvements
* Fullscreen mode (right-mouse button menu)
* Autozoom (disable via right-mouse button menu if it annoys)
* Toggle labels: right-mouse button menu, or press 'l' or 'L'
* Added 'oriented' attribute for edges: see e.g. Python/tree.py
* New example: Python/multiedge.py
* Ruby interface (rubigraph) by mootoh
* Higher rate of XMLRPC requests processed without reducing the framerate
* Improved error reporting; use ubigraph_server -quiet to suppress warnings
* Brought C API up to date
* API now guarantees that generated id's are in the range 0x40000000-0x7fffffff.
* Bug fix: jittery vertices caused by threading issues
* Bug fix: colors with uppercase hex digits 
* Bug fix: segfault due to non-threadsafe code on big graphs (#768202)
* Bug fix: style changes that affect layout now exit idle mode

Version alpha-0.2.1, 23 May 2008
* Built-in colour palette
* Automatic control of shapedetail to maintain frame rate
* Vertices now have 'visible' attribute
* Bug fix: jitter caused by bug in spline edges
* Bug fix: spacetree bug that caused stack overflows in some 
  circumstances (#768202)
* Bug fix: going idle while zooming

Version alpha-0.2, 22 May 2008
* Performance improvements
* Spline edges: set_edge_attribute(e, "spline", "true")
* Edge strokes: set_edge_attribute(e, "stroke", "dotted"/"dashed"/"solid")
* Higher rate of XMLRPC requests can be handled (i.e., graph can be
  updated faster).
* Bug fix (#104168,#270723,#356196): Edge labels, width work correctly
* Bug fix (#270099): Edge id's were not released when vertices removed



examples/Python/edges.py
import xmlrpclib

# Create an object to represent our server.
server_url = 'http://127.0.0.1:20738/RPC2'
server = xmlrpclib.Server(server_url)
G = server.ubigraph

G.clear()

y = G.new_vertex()

G.set_vertex_attribute(y, "color", "#ffff00")
G.set_vertex_attribute(y, "shape", "dodecahedron")
G.set_vertex_attribute(y, "size", "2.0")

tinyred = G.new_vertex_style(0)
greenedge = G.new_edge_style(0)

G.set_edge_style_attribute(greenedge, "color", "#00ff00")
G.set_edge_style_attribute(greenedge, "strength", "4.0")

for j in range(1,12):
  G.new_vertex_w_id(100+j)
  G.change_vertex_style(100+j,tinyred)
  ya = G.new_edge_w_id(200+j,y,100+j)

G.set_vertex_style_attribute(tinyred, "color", "#800000")
G.set_vertex_style_attribute(tinyred, "size", "0.2")
G.set_vertex_style_attribute(tinyred, "shape", "sphere")

G.set_edge_style_attribute(greenedge, "color", "#00ff00")
G.set_edge_style_attribute(greenedge, "strength", "4.0")



# Change some edge attributes

G.set_edge_attribute(201, "label", "labelled")
G.set_edge_attribute(202, "color", "#00ff00")
G.set_edge_attribute(202, "label", "green")

G.set_edge_attribute(203, "length", "3.0")
G.set_edge_attribute(203, "label", "length=3.0")

G.set_edge_attribute(204, "width", "2.0")
G.set_edge_attribute(204, "label", "width=2.0")

G.set_edge_attribute(205, "width", "4.0")
G.set_edge_attribute(205, "label", "width=4.0")

G.set_edge_attribute(206, "arrow", "true")
G.set_edge_attribute(206, "label", "arrow=true")

for j in range(1,5):
  e2 = G.new_edge(101, 102)
  G.set_edge_attribute(e2, "spline", "true")
  G.set_edge_attribute(e2, "label", "spline=true")

G.set_edge_attribute(208, "visible", "false")
G.set_vertex_attribute(108, "label", "visible=false")

G.set_edge_attribute(209, "fontsize", "24")
G.set_edge_attribute(209, "label", "fontsize=24")

G.set_edge_attribute(210, "fontcolor", "#ff0000")
G.set_edge_attribute(210, "label", "fontcolor=#ff0000")

G.set_edge_attribute(211, "stroke", "dashed")
G.set_edge_attribute(211, "label", "stroke=dashed")

e2 = G.new_edge(110, 111)
G.set_edge_attribute(e2, "stroke", "dotted")
G.set_edge_attribute(e2, "spline", "true")
G.set_edge_attribute(e2, "label", "stroke=dotted")

